File: lib/Ntreev.Windows.Forms.Grid/CellBase.cpp
//=====================================================================================================================
// Ntreev Grid for .Net 2.0.4478.19833
// https://github.com/NtreevSoft/GridControl
// 
// Released under the MIT License.
// 
// Copyright (c) 2010 Ntreev Soft co., Ltd.
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
// documentation files (the "Software"), to deal in the Software without restriction, including without limitation the 
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit 
// persons to whom the Software is furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the 
// Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE 
// WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR 
// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//=====================================================================================================================


#include "StdAfx.h"
#include "CellBase.h"
#include "GridControl.h"
#include "NativeClasses.h"

#include <vcclr.h>

namespace Ntreev { namespace Windows { namespace Forms { namespace Grid
{
    CellBase::CellBase(Ntreev::Windows::Forms::Grid::GridControl^ gridControl, GrCell* pCell)
        : GridObject(gridControl), m_pCell(pCell)
    {
        m_pCell->ManagedRef = this;
    }

    CellBase::CellBase(GrCell* pCell)
        : GridObject(pCell), m_pCell(pCell)
    {
        m_pCell->ManagedRef = this;
    }

    System::Drawing::Color CellBase::BackColor::get()
    {
        return m_pCell->GetBackColor();
    }

    void CellBase::BackColor::set(System::Drawing::Color color)
    {
        m_pCell->SetBackColor(color);
    }

    System::Drawing::Color CellBase::ForeColor::get()
    {
        return m_pCell->GetForeColor();
    }

    void CellBase::ForeColor::set(System::Drawing::Color color)
    {
        m_pCell->SetForeColor(color);
    }

    System::Drawing::Font^ CellBase::Font::get()
    {
        GrFont* pFont = (GrFont*)m_pCell->GetPaintingFont();
        return Native::WinFormFontManager::ToManagedFont(pFont);
    }

    void CellBase::Font::set(System::Drawing::Font^ font)
    {
        GrFont* pFont = Native::WinFormFontManager::FromManagedFont(font);
        m_pCell->SetFont(pFont);
    }

    bool CellBase::IsDisplayable::get()
    {
        return m_pCell->GetDisplayable();
    }

    void CellBase::ResetBackColor()
    {
        m_pCell->SetBackColor(GrColor::Empty);
    }

    void CellBase::ResetForeColor()
    {
        m_pCell->SetForeColor(GrColor::Empty);
    }

    bool CellBase::ShouldSerializeForeColor()
    {
        return m_pCell->GetForeColor(false) != GrColor::Empty;
    }

    bool CellBase::ShouldSerializeBackColor()
    {
        return m_pCell->GetBackColor(false) != GrColor::Empty;
    }

    bool CellBase::ShouldSerializeFont()
    {
        GrFont* pFont = m_pCell->GetFont(false);
        if(pFont == nullptr)
            return false;
        return pFont != this->GridCore->GetGridWindow()->GetDefaultFont();
    }

    bool CellBase::ShouldSerializePadding()
    {
        return m_pCell->GetPadding(false) != GrPadding::Default;
    }

    void CellBase::ResetPadding()
    {
        m_pCell->SetPadding(GrPadding::Default);
    }

    void CellBase::ResetFont()
    {
        m_pCell->SetFont(NULL);
    }

    System::Windows::Forms::Padding CellBase::Padding::get()
    {
        return m_pCell->GetPadding();
    }

    void CellBase::Padding::set(System::Windows::Forms::Padding value)
    {
        m_pCell->SetPadding(value); 
    }

    void CellBase::OnBackColorChanged(System::EventArgs^ e)
    {
        BackColorChanged(this, e);
    }

    void CellBase::OnForeColorChanged(System::EventArgs^ e)
    {
        ForeColorChanged(this, e);
    }

    void CellBase::OnFontChanged(System::EventArgs^ e)
    {
        FontChanged(this, e);
    }

    System::Drawing::Color CellBase::DisplayBackColor::get()
    {
        return m_pCell->GetPaintingBackColor();
    }

    System::Drawing::Color CellBase::DisplayForeColor::get()
    {
        return m_pCell->GetPaintingForeColor();
    }

    System::Drawing::Rectangle CellBase::ClientRectangle::get()
    {
        return m_pCell->GetClientRect();
    }

    System::Drawing::Rectangle CellBase::Bounds::get()
    {
        return m_pCell->GetRect();
    }

    int CellBase::X::get()
    {
        return m_pCell->GetX();
    }

    int CellBase::Y::get()
    {
        return m_pCell->GetY();
    }

    int CellBase::Width::get()
    {
        return m_pCell->GetWidth();
    }

    int CellBase::Height::get()
    {
        return m_pCell->GetHeight();
    }

    System::Drawing::Point CellBase::Location::get()
    {
        return m_pCell->GetLocation();
    }

    System::Drawing::Size CellBase::Size::get()
    {
        return m_pCell->GetSize();
    }

    int CellBase::Left::get()
    {
        return m_pCell->GetX();
    }

    int CellBase::Top::get()
    {
        return m_pCell->GetY();
    }

    int CellBase::Right::get()
    {
        return m_pCell->GetRight();
    }

    int CellBase::Bottom::get()
    {
        return m_pCell->GetBottom();
    }

    void CellBase::ForeColorChanged::add(System::EventHandler^ p)
    {
        m_eventForeColorChanged += p; 
    }

    void CellBase::ForeColorChanged::remove(System::EventHandler^ p)
    {
        m_eventForeColorChanged -= p; 
    }

    void CellBase::ForeColorChanged::raise(System::Object^ sender, System::EventArgs^ e)
    {
        if(m_eventForeColorChanged != nullptr) 
        {
            m_eventForeColorChanged->Invoke(sender, e); 
        } 
    }

    void CellBase::BackColorChanged::add(System::EventHandler^ p)
    {
        m_eventBackColorChanged += p; 
    }

    void CellBase::BackColorChanged::remove(System::EventHandler^ p)
    {
        m_eventBackColorChanged -= p; 
    }

    void CellBase::BackColorChanged::raise(System::Object^ sender, System::EventArgs^ e)
    {
        if(m_eventBackColorChanged != nullptr) 
        {
            m_eventBackColorChanged->Invoke(sender, e); 
        } 
    }

    void CellBase::FontChanged::add(System::EventHandler^ p)
    {
        m_eventFontChanged += p; 
    }

    void CellBase::FontChanged::remove(System::EventHandler^ p)
    {
        m_eventFontChanged -= p; 
    }

    void CellBase::FontChanged::raise(System::Object^ sender, System::EventArgs^ e)
    {
        if(m_eventFontChanged != nullptr) 
        {
            m_eventFontChanged->Invoke(sender, e); 
        } 
    }

    System::Drawing::Color CellBase::PaintingForeColor_ICellBase::get()
    {
        return m_pCell->GetPaintingForeColor();
    }

    System::Drawing::Color CellBase::PaintingBackColor_ICellBase::get()
    {
        return m_pCell->GetPaintingBackColor();
    }

    bool CellBase::Multiline_ICellBase::get()
    {
        return m_pCell->GetTextMulitiline();
    }

    bool CellBase::WordWrap_ICellBase::get()
    {
        return m_pCell->GetTextWordWrap();
    }

    System::Drawing::StringAlignment CellBase::Alignment_ICellBase::get()
    {
        return (System::Drawing::StringAlignment)m_pCell->GetTextHorzAlign();
    }

    System::Drawing::StringAlignment CellBase::LineAlignment_ICellBase::get()
    {
        return (System::Drawing::StringAlignment)m_pCell->GetTextVertAlign();
    }

    CellState CellBase::State_ICellBase::get()
    {
        GrFlag flag = m_pCell->ToPaintStyle();
        return (CellState)flag.Get((ulong)CellState::All);
    }
} /*namespace Grid*/ } /*namespace Forms*/ } /*namespace Windows*/ } /*namespace Ntreev*/

File: lib/WinformGridHost/CellBase.cs
using Ntreev.Library.Grid;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Ntreev.Windows.Forms.Grid
{
    public abstract class CellBase : GridObject, ICellBase
    {
        private readonly GrCell cell;

        internal CellBase(GridControl gridControl, GrCell cell)
            : base(gridControl)
        {
            this.cell = cell;
            this.cell.ManagedRef = this;
        }

        internal CellBase(GrCell pCell)
            : base(pCell)
        {
            this.cell = pCell;
            this.cell.ManagedRef = this;
        }

        public void ResetForeColor()
        {
            this.cell.SetForeColor(GrColor.Empty);
        }

        public void ResetBackColor()
        {
            this.cell.SetBackColor(GrColor.Empty);
        }

        public void ResetFont()
        {
            //this.pCell->SetFont(GrFont::Empty);
        }

        public void Invalidate()
        {
            this.cell.Invalidate();
        }

        [Description("셀의 전경색을 가져오거나 설정합니다.")]
        [Category("Appearance")]
        public Color ForeColor
        {
            get { return this.cell.GetForeColor(); }
            set { this.cell.SetForeColor(value); }
        }

        [Description("셀의 배경색을 가져오거나 설정합니다.")]
        [Category("Appearance")]
        public Color BackColor
        {
            get { return this.cell.GetBackColor(); }
            set { this.cell.SetBackColor(value); }
        }

        [Category("Appearance")]
        [AmbientValue((string)null)]
        public Font Font
        {
            get
            {
                GrFont pFont = this.cell.GetPaintingFont();
                return Utility.ToManaged(pFont);
            }
            set
            {
                GrFont pFont = Utility.FromManaged(value);
                this.cell.SetFont(pFont);
            }
        }

#if DEBUG
        [Category("Debug")]
#else
        [Browsable(false)]
#endif
        public bool IsDisplayable
        {
            get { return this.cell.IsDisplayable; }
        }

        [TypeConverter(typeof(StringConverter))]
        [Category("Data")]
        [DefaultValue((string)null)]
        public object Tag
        {
            get;
            set;
        }

#if DEBUG
        [Category("Debug")]
#else
        [Browsable(false)]
#endif
        public Rectangle Bounds
        {
            get { return this.cell.Bounds; }
        }

#if DEBUG
        [Category("Debug")]
#else
        [Browsable(false)]
#endif
        public int X
        {
            get { return this.cell.X; }
        }

#if DEBUG
        [Category("Debug")]
#else
        [Browsable(false)]
#endif
        public int Y
        {
            get { return this.cell.Y; }
        }

#if DEBUG
        [Category("Debug")]
#else
        [Browsable(false)]
#endif
        public virtual int Width
        {
            get { return this.cell.Width; }
            set
            {
                throw new NotImplementedException();
            }
        }

#if DEBUG
        [Category("Debug")]
#else
        [Browsable(false)]
#endif
        public virtual int Height
        {
            get { return this.cell.Height; }
            set
            {
                throw new NotImplementedException();
            }
        }

#if DEBUG
        [Category("Debug")]
#else
        [Browsable(false)]
#endif
        public Point Location
        {
            get { return this.cell.Location; }
        }

#if DEBUG
        [Category("Debug")]
#else
        [Browsable(false)]
#endif
        public Size Size
        {
            get { return this.cell.Size; }
        }

#if DEBUG
        [Category("Debug")]
#else
        [Browsable(false)]
#endif
        public int Left
        {
            get { return this.cell.X; }
        }

#if DEBUG
        [Category("Debug")]
#else
        [Browsable(false)]
#endif
        public int Top
        {
            get { return this.cell.Y; }
        }

#if DEBUG
        [Category("Debug")]
#else
        [Browsable(false)]
#endif
        public int Right
        {
            get { return this.cell.Right; }
        }

#if DEBUG
        [Category("Debug")]
#else
        [Browsable(false)]
#endif
        public int Bottom
        {
            get { return this.cell.Bottom; }
        }

#if DEBUG
        [Category("Debug")]
#else
		[Browsable(false)]
#endif
        public bool Capture
        {
            get { return this.cell.GetMousePressed(); }
        }

#if DEBUG
        [Category("Debug")]
#else
        [Browsable(false)]
#endif
        public Rectangle ClientRectangle
        {
            get { return this.cell.ClientRectangle; }
        }

        [Description("셀의 안쪽 여백을 가져오거나 설정합니다.")]
        [Category("Layout")]
        public Padding Padding
        {
            get { return this.cell.GetPadding(); }
            set { this.cell.SetPadding(value); }
        }



        internal Color DisplayForeColor
        {
            get { return this.cell.GetPaintingForeColor(); }
        }

        internal Color DisplayBackColor
        {
            get { return this.cell.GetPaintingBackColor(); }
        }

        internal GrCell NativeRef
        {
            get { return this.cell; }
        }


        private bool ShouldSerializeForeColor()
        {
            return this.cell.GetForeColorCore() != GrColor.Empty;
        }

        private bool ShouldSerializeBackColor()
        {
            return this.cell.GetBackColorCore() != GrColor.Empty;
        }

        private bool ShouldSerializeFont()
        {
            return this.cell.GetFontCore() != GrFont.Empty;
        }

        private bool ShouldSerializePadding()
        {
            return this.cell.GetPaddingCore() != GrPadding.Empty;
        }

        private void ResetPadding()
        {
            this.cell.SetPadding(GrPadding.Empty);
        }

        Color ICellBase.PaintingForeColor
        {
            get { return this.cell.GetPaintingForeColor(); }
        }

        Color ICellBase.PaintingBackColor
        {
            get { return this.cell.GetPaintingBackColor(); }
        }

        bool ICellBase.Multiline
        {
            get { return this.cell.IsTextMulitiline; }
        }

        bool ICellBase.WordWrap
        {
            get { return this.cell.IsTextWordWrap; }
        }

        StringAlignment ICellBase.Alignment
        {
            get { return (StringAlignment)this.cell.TextHorzAlign; }
        }

        StringAlignment ICellBase.LineAlignment
        {
            get { return (StringAlignment)this.cell.TextVertAlign; }
        }

        CellState ICellBase.State
        {
            get
            {
                GrPaintStyle flag = this.cell.ToPaintStyle();
                return (CellState)flag & CellState.All;
            }
        }
    }
}
